### ISSUE TITLE
regression: Type of "<name>" could not be determined because it refers to itself

### ISSUE THREAD

**Issue by imgurbot12** [microsoft/pyright#11685]
**Describe the bug**

Valid variable construction in real python code is now raising errors related to overlapping namespaces.

**Code or Screenshots**

```python
import queue

class Foo:
    queue: queue.Queue # Type of "queue" could not be determined because it refers to itself

    def __init__(self, queue: queue.Queue): # Variable not allowed in type expression
        self.queue = queue

foo = Foo(queue.Queue())
print(foo.queue)
```

Produces the following errors:

```
./test.py
  ./test.py:4:12 - error: Type of "queue" could not be determined because it refers to itself (reportGeneralTypeIssues)
 ./test.py:4:12 - error: Variable not allowed in type expression (reportInvalidTypeForm)
  ./test.py:6:31 - error: Variable not allowed in type expression (reportInvalidTypeForm)
3 errors, 0 warnings, 0 informations
```

**VS Code extension or command-line**
pyright 1.1.413 command-line and lsp.

**Comment by rchiodo** [comment]
Thanks for the issue but I'm not getting the same results:

<img width="897" height="351" alt="Image" src="https://github.com/user-attachments/assets/538ffd26-0937-4578-9cb6-2fbd504756ec" />

And in the pyright playground:

https://pyright-play.net/?strict=true&code=JYWwDg9gTgLgBARwK4FNUCh0GMA2BDAZwLgDEIIAudOGxVVCutFAOgEV6U4BiOAFQCeYLhABmcAETJmEuFghIcAEzgA7CPABGXJShgooIYKpQrtWPEgJdg8KClEHiMCHFvWcozLTi7xAfX9jW0CACg9RABomBhjWDmYASkZeADU8KGA8TRwudXg8HBwIAHdTN1U4GCEuFAAPMHsiYAhVah9aCJZpVDgAXjjMUXJ%2B0nJQnvjOUMTE9EbjGFDhiG7OOaA

Same output I see. 

Do you have your own `Queue`?

**Comment by imgurbot12** [comment]
@rchiodo thanks for following up, but I don't understand how we're getting this discrepancy.

Is the playground running a different version than the ones published?
I've installed pyright via npm and the pip package both advertised on the official wiki and both come up with the same error.
Is there some cache or something that might be causing a problem?

I do not have any queue library installed so far as I know. its not sitting next to any queue.py file either.

**Comment by rchiodo** [comment]
If you go to def on `queue.Queue`, where does it go? 

There's no cache that would cause this. The only thing I know of is if we're finding a different Queue for some reason.

For me it comes out here:

<img width="803" height="364" alt="Image" src="https://github.com/user-attachments/assets/5d5ed323-290c-4bb6-a8d3-b8b1e7351384" />

**Comment by rchiodo** [comment]
Oh wait, if I switch Python 3.10 I get the same error as you. What Python version are you using?

**Comment by rchiodo** [comment]
Yeah I switch the playground to 3.13 or earlier I get the same error output as you. Which I believe is by design or at least not a regression as it reproduces with every Pyright version I tried.

**Comment by rchiodo** [comment]
Ah actually it was this that is the difference: https://peps.python.org/pep-0749/. In 3.14 and above Pyright parses the code differently because of that Pep.

So this error is expected if you're using 3.13 or below.

**Comment by imgurbot12** [comment]
@rchiodo
> So this error is expected if you're using 3.13 or below.

I've been using pyright for many years at this point and this is the very first time I've encountered this since I started using it. This has to have been a recent change to pyright. Are you saying this is intended behavior? That seems silly. The code is still valid in older versions as well so I don't understand the reason why it would error at valid python code.

Interestingly, if I install pyright via pip using python 3.14 it does not have any errors, but using the version available via npm still raises the same errors.

**Comment by rchiodo** [comment]
Try it with older versions, same error. 

It's specifically this:

```
import queue 

class Foo(): 
    queue: queue.Queue
```

As discussed here: https://peps.python.org/pep-0649/#comparison-of-annotation-semantics, the order of how annotations are parsed changed in 3.14.

**Comment by rchiodo** [comment]
This might have always been a bug in Pyright, but I'd suggest switching to 3.14 (or don't name variables the same as package names)

This fixes it too (in 3.13 and below):

```
import queue

class Foo:
    q: queue.Queue 

    def __init__(self, q: queue.Queue): 
        self.q = q

foo = Foo(queue.Queue())
print(foo.q)
```

**Comment by rchiodo** [comment]
Yeah I can reproduce this with the oldest version we have in the Playground:

https://pyright-play.net/?pyrightVersion=1.1.364&pythonVersion=3.13&code=JYWwDg9gTgLgBARwK4FNUCh0GMA2BDAZwLgDEIIAudOGxVVCutFAOgEV6U4BiOAFQCeYLhABmcAETJmEuFghIcAEzgA7CPABGXJShgooIYKpQrtWPEgJdg8KClEHiMCHFvWcozLTi7xAfX9jW0CACg9RABomBhjWDmYASkZeADU8KGA8TRwudXg8HBwIAHdTN1U4GCEuFAAPMHsiYAhVah9aCJZpVDgAXjjMUXJ%2B0nJQnvjOUMTE9EbjGFDhiG7OOaA

**Comment by imgurbot12** [comment]
I appreciate the thorough follow up thanks. I rolled back a few times to check myself and it does reproduce consistently as you mention.

In all my years using python I would've sworn I'd never had these issues before but maybe I magically never ran into this that I can remember. I guess I'll just stick with 3.14. Thanks again.
